autopush/senderids.py

```diff
@@ -40,63 +41,73 @@
 
 class SenderIDs(object):
     """Handle Read, Write and cache of SenderID values from S3"""
-    _updated = 0
     _expry = SENDERID_EXPRY
     _senderIDs = {}
     _use_s3 = True
     KEYNAME = "senderids"
+    service = None
 
     def __init__(self, args):
         """Optionally load or fetch the set of SenderIDs from S3"""
         self.conn = S3Connection()
-        self.ID = args.get("s3_bucket", DEFAULT_BUCKET)
+        self.ID = args.get("s3_bucket", DEFAULT_BUCKET).lower()
         self._expry = args.get("senderid_expry", SENDERID_EXPRY)
         self._use_s3 = args.get("use_s3", True)
         senderIDs = args.get("senderid_list", {})
+        self.service = None
         if senderIDs:
             if type(senderIDs) is not dict:
                 log.err("senderid_list is not a dict. Ignoring")
             else:
+                # We're initializing, so it's ok to block.
                 self.update(senderIDs)
-        self._refresh()
 
-    def _write(self, bucket, senderIDs):
+    def start(self):
+        if self._use_s3:
+            log.msg("Starting SenderID service...")
+            self.service = TimerService(self._expry, self._refresh)
+            self.service.startService()
+
+    def _write(self, senderIDs, *args):
         """Write a list of SenderIDs to S3"""
+        bucket = self.conn.get_bucket(self.ID)
         key = Key(bucket)
         key.key = self.KEYNAME
         key.set_contents_from_string(json.dumps(senderIDs))
         self._senderIDs = senderIDs
-        self._updated = time.time()
 
-    def _create(self, senderIDs):
+    def _err(self, state):
+        if (isinstance(state.value, S3ResponseError) and
+                state.value.reason == 'Not Found'):
+            self._create()
+
+    def _create(self, *args):
         """Create a new bucket containing the senderIDs"""
-        bucket = self.conn.create_bucket(self.ID)
-        self._write(bucket, senderIDs)
+        self.conn.create_bucket(self.ID)
+        self._write(self._senderIDs)
+
+    def _update_senderIDs(self, *args):
+        bucket = self.conn.get_bucket(self.ID)
+        key = Key(bucket)
+        key.key = self.KEYNAME
+        candidates = json.loads(key.get_contents_as_string())
+        if candidates:
+            if type(candidates) is not dict:
+                log.err("Wrong data type stored for senderIDs. "
+                        "Should be dict. Ignoring.")
+                return
+            self._senderIDs = candidates
```

Comment:
This isn't thread-safe. I think you can just return `candidates` here. Then, in `_refresh`, you can write `d.addCallback(self._set_senderIDs)`.